File: data-structures/ds2-linked-list/linked-list/suxiong-python/linked_list_test.py
import pytest
from linked_list import LinkedList

@pytest.fixture
def linked_list():
    return LinkedList()

def test_unit_add_first(linked_list):
    linked_list.add_first("First")
    assert linked_list.get_first() == "First"

def test_unit_add_last(linked_list):
    linked_list.add_last("Last")
    assert linked_list.get_last() == "Last"

def test_unit_add(linked_list):
    linked_list.add("Middle", 0)
    assert linked_list.get(0) == "Middle"

def test_unit_remove(linked_list):
    linked_list.add_first("First")
    linked_list.add_last("Last")
    assert linked_list.remove(0) == "First"

def test_unit_get_size(linked_list):
    linked_list.add_first("First")
    linked_list.add_last("Last")
    assert linked_list.get_size() == 2

def test_integration_add_and_remove(linked_list):
    linked_list.add_first("First")
    linked_list.add_last("Last")
    linked_list.add("Middle", 1)
    assert linked_list.remove(0) == "First"
    assert linked_list.remove(0) == "Middle"
    assert linked_list.remove(0) == "Last"

def test_integration_add_and_get(linked_list):
    linked_list.add_first("First")
    linked_list.add_last("Last")
    linked_list.add("Middle", 1)
    assert linked_list.get(0) == "First"
    assert linked_list.get(1) == "Middle"
    assert linked_list.get(2) == "Last"

def test_integration_get_size(linked_list):
    linked_list.add_first("First")
    linked_list.add_last("Last")
    linked_list.add("Middle", 1)
    assert linked_list.get_size() == 3

def test_integration_is_empty(linked_list):
    assert linked_list.get_size() == 0
    linked_list.add_first("First")
    assert linked_list.get_size() != 0

def test_integration_add_and_set_and_remove(linked_list):
    linked_list.add_first("First")
    linked_list.add_last("Last")
    linked_list.add("Middle", 1)
    assert linked_list.get_size() == 3

    linked_list.set_node("NewFirst", 0)
    linked_list.set_node("NewMiddle", 1)
    linked_list.set_node("NewLast", 2)
    assert linked_list.get_size() == 3

    assert linked_list.remove_first() == "NewFirst"
    assert linked_list.get_size() == 2

    assert linked_list.remove_first() == "NewMiddle"
    assert linked_list.get_size() == 1

    assert linked_list.remove_first() == "NewLast"
    assert linked_list.get_size() == 0

def test_integration_add_and_set_and_remove_last(linked_list):
    linked_list.add_first("First")
    linked_list.add_last("Last")
    linked_list.add("Middle", 1)
    assert linked_list.get_size() == 3

    linked_list.set_node("NewFirst", 0)
    linked_list.set_node("NewMiddle", 1)
    linked_list.set_node("NewLast", 2)
    assert linked_list.get_size() == 3

    assert linked_list.remove_last() == "NewLast"
    assert linked_list.get_size() == 2

    assert linked_list.remove_last() == "NewMiddle"
    assert linked_list.get_size() == 1

    assert linked_list.remove_last() == "NewFirst"
    assert linked_list.get_size() == 0

def test_integration_add_and_set_and_remove_middle(linked_list):
    linked_list.add_first("First")
    linked_list.add_last("Last")
    linked_list.add("Middle", 1)
    assert linked_list.get_size() == 3

    linked_list.set_node("NewFirst", 0)
    linked_list.set_node("NewMiddle", 1)
    linked_list.set_node("NewLast", 2)
    assert linked_list.get_size() == 3

    assert linked_list.remove(1) == "NewMiddle"
    assert linked_list.get_size() == 2

    assert linked_list.remove(1) == "NewLast"
    assert linked_list.get_size() == 1

    assert linked_list.remove(0) == "NewFirst"
    assert linked_list.get_size() == 0

def test_integration_add_and_set_and_remove_out_of_bounds(linked_list):
    linked_list.add_first("First")
    linked_list.add_last("Last")
    linked_list.add("Middle", 1)
    assert linked_list.get_size() == 3

    linked_list.set_node("NewFirst", 0)
    linked_list.set_node("NewMiddle", 1)
    linked_list.set_node("NewLast", 2)
    assert linked_list.get_size() == 3

    with pytest.raises(IndexError):
        linked_list.remove(3)
    assert linked_list.get_size() == 3

    with pytest.raises(IndexError):
        linked_list.remove(-1)
    assert linked_list.get_size() == 3

def test_integration_add_and_set_and_remove_empty(linked_list):
    assert linked_list.get_size() == 0

    with pytest.raises(IndexError):
        linked_list.remove_first()
    assert linked_list.get_size() == 0

    with pytest.raises(IndexError):
        linked_list.remove_last()
    assert linked_list.get_size() == 0

    with pytest.raises(IndexError):
        linked_list.remove(0)
    assert linked_list.get_size() == 0

def test_integration_add_and_set_and_remove_one(linked_list):
    linked_list.add_first("First")
    assert linked_list.get_size() == 1

    assert linked_list.remove_first() == "First"
    assert linked_list.get_size() == 0

    linked_list.add_last("Last")
    assert linked_list.get_size() == 1

    assert linked_list.remove_last() == "Last"
    assert linked_list.get_size() == 0

    linked_list.add_first("First")
    assert linked_list.get_size() == 1

    assert linked_list.remove(0) == "First"
    assert linked_list.get_size() == 0

File: data-structures/ds2-linked-list/linked-list/javascript/linkedListTest.ts
import { LL } from "./linkedList";

describe("LinkedList", () => {
  let linkedList: LL<string>;

  beforeEach(() => {
    linkedList = new LL<string>();
  });

  test("Unit: add first", () => {
    linkedList.addFirst("First");
    expect(linkedList.getFirst()).toBe("First");
  });

  test("Unit: add last", () => {
    linkedList.addLast("Last");
    expect(linkedList.getLast()).toBe("Last");
  });

  test("Unit: add", () => {
    linkedList.add("Middle", 0);
    expect(linkedList.get(0)).toBe("Middle");
  });

  test("Unit: remove", () => {
    linkedList.addFirst("First");
    linkedList.addLast("Last");
    expect(linkedList.remove(0)).toBe("First");
  });

  test("Unit: get size", () => {
    linkedList.addFirst("First");
    linkedList.addLast("Last");
    expect(linkedList.getSize()).toBe(2);
  });

  test("Integration: add and remove", () => {
    linkedList.addFirst("First");
    linkedList.addLast("Last");
    linkedList.add("Middle", 1);
    expect(linkedList.remove(0)).toBe("First");
    expect(linkedList.remove(0)).toBe("Middle");
    expect(linkedList.remove(0)).toBe("Last");
  });

  test("Integration: add and get", () => {
    linkedList.addFirst("First");
    linkedList.addLast("Last");
    linkedList.add("Middle", 1);
    expect(linkedList.get(0)).toBe("First");
    expect(linkedList.get(1)).toBe("Middle");
    expect(linkedList.get(2)).toBe("Last");
  });

  test("Integration: get size", () => {
    linkedList.addFirst("First");
    linkedList.addLast("Last");
    linkedList.add("Middle", 1);
    expect(linkedList.getSize()).toBe(3);
  });

  test("Integration: is empty", () => {
    expect(linkedList.getSize()).toBe(0);
    linkedList.addFirst("First");
    expect(linkedList.getSize()).not.toBe(0);
  });

  test("Integration: add and set and remove", () => {
    linkedList.addFirst("First");
    linkedList.addLast("Last");
    linkedList.add("Middle", 1);
    expect(linkedList.getSize()).toBe(3);

    linkedList.set("NewFirst", 0);
    linkedList.set("NewMiddle", 1);
    linkedList.set("NewLast", 2);
    expect(linkedList.getSize()).toBe(3);

    expect(linkedList.removeFirst()).toBe("NewFirst");
    expect(linkedList.getSize()).toBe(2);

    expect(linkedList.removeFirst()).toBe("NewMiddle");
    expect(linkedList.getSize()).toBe(1);

    expect(linkedList.removeFirst()).toBe("NewLast");
    expect(linkedList.getSize()).toBe(0);
  });

  test("Integration: add and set and remove last", () => {
    linkedList.addFirst("First");
    linkedList.addLast("Last");
    linkedList.add("Middle", 1);
    expect(linkedList.getSize()).toBe(3);

    linkedList.set("NewFirst", 0);
    linkedList.set("NewMiddle", 1);
    linkedList.set("NewLast", 2);
    expect(linkedList.getSize()).toBe(3);

    expect(linkedList.removeLast()).toBe("NewLast");
    expect(linkedList.getSize()).toBe(2);

    expect(linkedList.removeLast()).toBe("NewMiddle");
    expect(linkedList.getSize()).toBe(1);

    expect(linkedList.removeLast()).toBe("NewFirst");
    expect(linkedList.getSize()).toBe(0);
  });

  test("Integration: add and set and remove middle", () => {
    linkedList.addFirst("First");
    linkedList.addLast("Last");
    linkedList.add("Middle", 1);
    expect(linkedList.getSize()).toBe(3);

    linkedList.set("NewFirst", 0);
    linkedList.set("NewMiddle", 1);
    linkedList.set("NewLast", 2);
    expect(linkedList.getSize()).toBe(3);

    expect(linkedList.remove(1)).toBe("NewMiddle");
    expect(linkedList.getSize()).toBe(2);

    expect(linkedList.remove(1)).toBe("NewLast");
    expect(linkedList.getSize()).toBe(1);

    expect(linkedList.remove(0)).toBe("NewFirst");
    expect(linkedList.getSize()).toBe(0);
  });

  test("Integration: add and set and remove out of bounds", () => {
    linkedList.addFirst("First");
    linkedList.addLast("Last");
    linkedList.add("Middle", 1);
    expect(linkedList.getSize()).toBe(3);

    linkedList.set("NewFirst", 0);
    linkedList.set("NewMiddle", 1);
    linkedList.set("NewLast", 2);
    expect(linkedList.getSize()).toBe(3);

    expect(() => linkedList.remove(3)).toThrowError(RangeError);
    expect(linkedList.getSize()).toBe(3);

    expect(() => linkedList.remove(-1)).toThrowError(RangeError);
    expect(linkedList.getSize()).toBe(3);
  });

  test("Integration: add and set and remove empty", () => {
    expect(linkedList.getSize()).toBe(0);

    expect(() => linkedList.removeFirst()).toThrowError(RangeError);
    expect(linkedList.getSize()).toBe(0);

    expect(() => linkedList.removeLast()).toThrowError(RangeError);
    expect(linkedList.getSize()).toBe(0);

    expect(() => linkedList.remove(0)).toThrowError(RangeError);
    expect(linkedList.getSize()).toBe(0);
  });

  test("Integration: add and set and remove one", () => {
    linkedList.addFirst("First");
    expect(linkedList.getSize()).toBe(1);

    expect(linkedList.removeFirst()).toBe("First");
    expect(linkedList.getSize()).toBe(0);

    linkedList.addLast("Last");
    expect(linkedList.getSize()).toBe(1);

    expect(linkedList.removeLast()).toBe("Last");
    expect(linkedList.getSize()).toBe(0);

    linkedList.addFirst("First");
    expect(linkedList.getSize()).toBe(1);

    expect(linkedList.remove(0)).toBe("First");
    expect(linkedList.getSize()).toBe(0);
  });
});
